### PR TITLE
Align over multiple columns

### PR DESCRIPTION
**Pull request by Strackeror**
I added support for aligning over multiple columns, mimicking the behavior of kakoune when having more than one selection per line.

### CODE REVIEW

**Review by 71 — CHANGES_REQUESTED** [review]
That's a great feature to have, thanks! On top of little nits, I have some feedback:
1. No matter which approach we end up going for, please add tests. Especially tests with multiple selections / line, lines with different amounts of selections, ...
2. Calling `edit` in a loop isn't great IMO, on a slow computer it might look weird and cause issues if the user is editing the document at the same time. Ideally you would compute all columns you will need, and then for each selection figure out what the edit is to make it respect the corresponding column.
   * AFAIK the first column is (roughly) `max(...firstSelectionOfEachLine)`, and subsequent columns are `previousColumn + max(...distanceBetweenPreviousColumnAndNextColumnOfEachLine)`.

**Comment by Strackeror** [comment]
Alright, thank you for the feedback !

The lack of test is... irresponsible of me, I'm generally the sole user of my personal projects, so I tend to just do some manual tests and leave it there, I'll get on writing some !

I also have a prototype which worked within one edit block, but I had to keep track of how many characters were added to each line, as the selections got out of sync. I'll switch to that one.

**Comment by 71** [comment]
Haha, for a while Dance barely had any tests, but a lot of commands have edge cases so I eventually had to add them, and I added a lot of helpers to make my life as easy as possible writing tests. Hopefully you'll figure things out easily (there should be a small tutorial in the test directory README), and with them and a few test cases you should be confident your code will work (even after having to keep track of changes by yourself and stuff like that).

**Comment by Strackeror** [comment]
So, I wrote some basic tests, and matched behavior on kakoune/helix on a couple points:
- Aligning is now done based on the selection active cursor instead of selection start
- We straight up stop the alignment process if any selection is on multiple lines
I ended up investigating the exact behavior when writing the tests and noticed the discrepancies, another point in favor of writing tests I suppose :)

Finally, I also moved the whole code within a single edit block, which required completely changing the logic.

File: test/suite/commands/edit-align.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     ^^^ 1
+ --> 99) dolor
+     ^^^ 2
+ --> 100) sit
+     ^^^^ 3
+ --> 101) amet
+     ^^^^ 4
+```
+
+## 2 align
```
Comment (71) [inline test/suite/commands/edit-align.md:43]: I'd like to have a test that also shows what happens when selections are reversed (so nothing changes), to further display the behavior.

File: test/suite/commands/edit-align.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                       ^^^^^ 2
+consectetur adipiscing  elit Morbi eget
+^^^^^^^^^^^ 3
+            ^^^^^^^^^^ 4
+                        ^^^^ 5
+                             ^^^^^ 6
+                                   ^^^^ 7
+    Aliquam       erat
+    ^^^^^^^ 8
+                  ^^^^ 9
```
Comment (71) [inline test/suite/commands/edit-align.md:99]: FYI you can have multiple "selection indicators" per line, e.g.
```suggestion
      Lorem      ipsum dolor
      ^^^^^ 0          ^^^^^ 2
                 ^^^^^ 1
consectetur adipiscing  elit Morbi eget
^^^^^^^^^^^ 3           ^^^^ 5     ^^^^ 7
            ^^^^^^^^^^ 4     ^^^^^ 6
    Aliquam       erat
    ^^^^^^^ 8     ^^^^ 9
```

File: test/suite/commands/edit-align.md (hunk 1 above)
Comment (71) [reply test/suite/commands/edit-align.md:43]: Ha, never mind, you did this below, thanks!

File: test/suite/commands/edit-align.md
Comment (71) [inline test/suite/commands/edit-align.md:1]: Thanks for the test, it really shows how much more powerful the command is now!

File: src/commands/edit.ts [hunk 3]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
+    const selectionByColumn: vscode.Selection[][] = [];
+    let currentLine: number | undefined = undefined;
+    let currentColumn: number = 0;
+    for (const selection of sortedSelections) {
+      if (selection.start.line !== selection.end.line) {
+        throw new Error(
+          "Cannot align selections that are spanning multiple lines",
```
Comment (71) [inline src/commands/edit.ts:296]: Nice to be handling this edge case now! 😅

```suggestion
          "cannot align selections that span multiple lines",
```

To be more consistent with other error messages.

File: src/commands/edit.ts [hunk 4]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
-    for (let i = 0, len = selections.length; i < len; i++) {
-      const selection = selections[i];
+      if (!(currentColumn in selectionByColumn)) {
+        selectionByColumn[currentColumn] = [];
+      }
 
-      builder.insert(selection.start, fill.repeat(startChar - selection.start.character));
+      selectionByColumn[currentColumn].push(selection);
+      currentColumn += 1;
+    }
```
Comment (71) [inline src/commands/edit.ts:311]:
```suggestion
    }

```

File: src/commands/edit.ts [hunk 5]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
  */
-export function align(
-  _: Context,
-  selections: readonly vscode.Selection[],
+export function align(_: Context, fill: Argument<string> = " ") {
+  return edit((builder, selections) => {
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
+    const selectionByColumn: vscode.Selection[][] = [];
```
Comment (71) [inline src/commands/edit.ts:290]:
```suggestion
    const selectionsByColumn: vscode.Selection[][] = [];
```

File: src/commands/edit.ts [hunk 6]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
-  return edit((builder, selections) => {
-    for (let i = 0, len = selections.length; i < len; i++) {
-      const selection = selections[i];
+      if (!(currentColumn in selectionByColumn)) {
+        selectionByColumn[currentColumn] = [];
+      }
 
-      builder.insert(selection.start, fill.repeat(startChar - selection.start.character));
+      selectionByColumn[currentColumn].push(selection);
+      currentColumn += 1;
```
Comment (71) [inline src/commands/edit.ts:310]:
```suggestion
      currentColumn++;
```

File: src/commands/edit.ts [hunk 7]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+    }
+    // Selections aren't updated as we fill each line, so we keep track of how
+    // many characters we added to each line as we go.
+    const lineFillCounters = new Map();
```
Comment (71) [inline src/commands/edit.ts:314]:
```suggestion
    const lineFillCounters = new Map<number, number>();
```

File: src/commands/edit.ts [hunk 8]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+        currentLine = selection.start.line;
+        currentColumn = 0;
+      }
 
-  return edit((builder, selections) => {
-    for (let i = 0, len = selections.length; i < len; i++) {
-      const selection = selections[i];
+      if (!(currentColumn in selectionByColumn)) {
+        selectionByColumn[currentColumn] = [];
+      }
```
Comment (71) [inline src/commands/edit.ts:307]: IIUC, `currentColumn` always increases by one, so this check can be replaced by:

```suggestion
      if (currentColumn === selectionsByColumn.length) {
        selectionsByColumn.push([]);
      }
```

File: src/commands/edit.ts [hunk 9]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
  * @keys `&` (kakoune: normal)
  */
-export function align(
-  _: Context,
-  selections: readonly vscode.Selection[],
+export function align(_: Context, fill: Argument<string> = " ") {
+  return edit((builder, selections) => {
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
```
Comment (71) [inline src/commands/edit.ts:289]: I was a bit confused at first, since "the nth column" often means "a line's nth character" in the context of a text editor. Perhaps `selectionsToAlign` or `selectionGroups` might be more understandable? Otherwise, please emphasize that "column" here is different from the typical meaning of "column" in a text editor in the comment.

File: src/commands/edit.ts [hunk 10]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+    const selectionByColumn: vscode.Selection[][] = [];
+    let currentLine: number | undefined = undefined;
```
Comment (71) [inline src/commands/edit.ts:291]: To avoid having `currentLine` change types, maybe we could initialize it as `0`, since it results as a no-op in the condition that checks `selection.start.line !== currentLine` below.

File: src/commands/edit.ts [hunk 11]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
 
-      builder.insert(selection.start, fill.repeat(startChar - selection.start.character));
+      selectionByColumn[currentColumn].push(selection);
+      currentColumn += 1;
+    }
+    // Selections aren't updated as we fill each line, so we keep track of how
+    // many characters we added to each line as we go.
+    const lineFillCounters = new Map();
+    const getAlignChar = (sel: vscode.Selection) =>
+      sel.active.character + (lineFillCounters.get(sel.start.line) ?? 0);
```
Comment (71) [inline src/commands/edit.ts:316]: Reaaaally small nit: I'd prefer using `active` instead of `start` here, to avoid asking yourself "why are these values different?" and remembering that the line is always the same when reading this code.

File: src/commands/edit.ts [hunk 12]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
-  _: Context,
-  selections: readonly vscode.Selection[],
+export function align(_: Context, fill: Argument<string> = " ") {
+  return edit((builder, selections) => {
+    const sortedSelections = Selections.sort(Direction.Forward, [...selections]);
+
+    // Group selections by 'column', nth column being nth selections of each line.
+    const selectionByColumn: vscode.Selection[][] = [];
+    let currentLine: number | undefined = undefined;
+    let currentColumn: number = 0;
```
Comment (71) [inline src/commands/edit.ts:292]:
```suggestion
    let currentColumn = 0;
```

File: src/commands/edit.ts [hunk 13]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+      selectionByColumn[currentColumn].push(selection);
+      currentColumn += 1;
+    }
+    // Selections aren't updated as we fill each line, so we keep track of how
+    // many characters we added to each line as we go.
+    const lineFillCounters = new Map();
+    const getAlignChar = (sel: vscode.Selection) =>
+      sel.active.character + (lineFillCounters.get(sel.start.line) ?? 0);
+
+    for (const selections of selectionByColumn) {
```
Comment (71) [inline src/commands/edit.ts:318]:
```suggestion
    for (const selectionsInColumn of selectionsByColumn) {
```

(Or whatever we decide to choose `selectionsByColumn`)

IMO this helps clarify what this list of selections has in common, and what the point of the further processing is (without having to read the enclosing context).

File: src/commands/edit.ts [hunk 14]
```diff
@@ -282,22 +282,50 @@ export async function replaceCharacters(
...
+      const furthestChar = Math.max(...selections.map(getAlignChar));
+      for (const selection of selections) {
+        const addCount = furthestChar - getAlignChar(selection);
+        builder.insert(selection.start, fill.repeat(addCount));
```
Comment (71) [inline src/commands/edit.ts:322]: I'm kinda curious. In Kakoune and Helix, did you check if there is a way to change the `fill` character? If so, what happens when it's multiple characters long? It doesn't matter that much, just curious.

**Review by 71 — CHANGES_REQUESTED** [review]
Great work!

Sorry, I have a ton of nits because your code is already very good, so I tried to make it as close to perfect as I could (at least subjectively, and in the context of other Dance code).

File: test/suite/commands/edit-align.md (hunk 2 above)
Comment (Strackeror) [reply test/suite/commands/edit-align.md:99]: Noooo, It was so annoying to set them up, that would have been so much easier D:

File: src/commands/edit.ts (hunk 14 above)
Comment (71) [reply src/commands/edit.ts:324]: Ah, I meant checking the width of the text in bytes (e.g. if `fill` is `abc`), but it doesn't matter that much, I don't think a feature would often be useful (or maybe it would be used with `0` or a similar 1-character string).

**Comment by tshort** [comment]
I'd just like to add my thanks for this feature. Great work! It'll help me.

File: src/commands/edit.ts (hunk 10 above)
Comment (71) [reply src/commands/edit.ts]: Sure, sounds good to me as well. `0` wasn't meant to mean "cannot be a line", but AFAIK things work normally if you start at `0` as well. Doesn't matter that much.

**Review by 71 — APPROVED** [review]
Fantastic, thank you!
